### PR TITLE
gadget,overlord: do not try to update if no content is defined

### PR DESCRIPTION
**Pull request by alfonsosanchezbeato**
Make sure to avoid additional progress in the moment we know an update
will not happen as no content is defined in a structure, so we do not
need to look for mount points for such structures, etc. This change
forces to make some changes in tests to make them more
realistic (content needs to be provided to reach some code paths while
this was not necessary before this change). Fixes LP#2035197.

### CODE REVIEW

**Review by andrewphelpsj — APPROVED** [review]
Looks good, thanks!

File: gadget/update.go [hunk 1]
```diff
@@ -1620,8 +1620,8 @@ func resolveUpdate(oldVol *PartiallyLaidOutVolume, newVol *LaidOutVolume, policy
 			if err != nil {
 				return nil, err
 			}
-			// Nothing to do after filtering
-			if filter != nil && len(resolvedContent) == 0 && len(newStruct.LaidOutContent) == 0 {
+			// No resolved or raw content that would need updating
+			if len(resolvedContent) == 0 && len(newStruct.LaidOutContent) == 0 {
 				continue
```
Comment (pedronis) [inline gadget/update.go:1625]: there's no unit test that hits the continue now, maybe there wasn't any even on master? anyway we might want one

File: gadget/update.go [hunk 2]
```diff
@@ -1620,8 +1620,8 @@ func resolveUpdate(oldVol *PartiallyLaidOutVolume, newVol *LaidOutVolume, policy
 			if err != nil {
 				return nil, err
 			}
-			// Nothing to do after filtering
-			if filter != nil && len(resolvedContent) == 0 && len(newStruct.LaidOutContent) == 0 {
```
Comment (pedronis) [inline gadget/update.go:1624]: do you know why the check was like this? would there be a reason to go over things even if there's no content?

File: tests/nested/manual/remodel-uc20-to-uc22/task.yaml [hunk 3]
```diff
@@ -21,20 +30,23 @@ environment:
   # vault
   SERIAL: 1234-snapd-remodel-minsize
   GADGET_NAME: test-snapd-remodel-pc
-  NEW_GADGET_NAME: test-snapd-remodel-pc-min-size
+  # use new gadget with shim signed by snakeoil keys so encrypted case works
+  NEW_GADGET_NAME: test-snapd-remodel-pc-snakeoil
```
Comment (pedronis) [inline tests/nested/manual/remodel-uc20-to-uc22/task.yaml:34]: I'm not sure I remember where is this new gadget defined/coming from?

File: gadget/update.go (hunk 1 above)
Comment (alfonsosanchezbeato) [reply gadget/update.go:1625]: There are manager tests that go through this, I have changed a test in gadget to run this nonetheless.

File: tests/nested/manual/remodel-uc20-to-uc22/task.yaml (hunk 3 above)
Comment (alfonsosanchezbeato) [reply tests/nested/manual/remodel-uc20-to-uc22/task.yaml:36]: This is a gadget packed from tests/lib/snaps/test-snapd-remodel-pc-min-size-22/, but with shim re-signed. I've added a comment to clarify.

**Review by alfonsosanchezbeato — COMMENTED** [review]
@pedronis thanks for the review, I've addressed your comments now

File: gadget/update_test.go [hunk 4]
```diff
@@ -1032,18 +1041,22 @@ func (u *updateTestSuite) TestUpdateApplyUC20MissingInitialMapFullLogicOnlySyste
...
+	// mbr - bare structure, and bios - partition w/o filesystem
+	for i, imgName := range []string{"mbr.img", "bios.img"} {
+		imgPath := filepath.Join(newData.RootDir, imgName)
+		err = os.WriteFile(imgPath, nil, 0644)
+		c.Assert(err, IsNil)
+		newData.Info.Volumes["pc"].Structure[i].Content = []gadget.VolumeContent{{Image: imgName}}
+		newData.Info.Volumes["pc"].Structure[i].Update.Edition = 1
+	}
+	// ubuntu-{seed,boot,save,data}
+	for i, fName := range []string{"seed", "boot", "save", "data"} {
```
Comment (pedronis) [inline gadget/update_test.go:1053]: I understand that the test was like this before but it's a bit weird to put content in save/data, same also in some other tests

**Review by pedronis — APPROVED** [review]
thank you, remark about the tests (as they already were)

**Comment by alfonsosanchezbeato** [comment]
Thanks for the reviews, I've rebased now.

**Pull request merged**